<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1">



<title>Predicting maximal growth rates with gRodon</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />




</head>

<body>




<h1 class="title toc-ignore">Predicting maximal growth rates with <code>gRodon</code></h1>



<p>The <code>gRodon</code> package allows you to predict the maximal growth rate of a prokaryotic organism from genomic data. For details of how this is done please see the original paper. Briefly, gRodon exploits codon usage statistics to detect optimization of highly-expressed genes, an indicator of rapid growth. We will provide some guidance about when these predictions are appropriate below.</p>
<p>The <code>gRodon</code> package is quite simple, with only a single function currently available to users: <code>predictGrowth()</code>. We provide a simple, easy to use interface that allows any user with a genome in-hand to predict maximal growth rate. Note that we will use “maximal growth rate”, “minimal doubling time”, and “growth rate” interchangeably below when talking about the output from <code>gRodon</code>.</p>
<p>The <code>gRodon</code> approach is largely based on an earlier program, <code>growthpred</code>, developed by <a href="https://doi.org/10.1371/journal.pgen.1000808">Viera-Silva et al.</a>. <code>gRodon</code> is intended to be:<br />
1. More user-friendly (especially for R users)<br />
2. More accurate (due to the incorporation of additional measures of codon usage)</p>
<div id="before-you-begin" class="section level2">
<h2>Before You Begin</h2>
<p>There are two things that <code>gRodon</code> does <em>not</em> do that you <em>must</em> do ahead of time yourself in order to predict growth rate:<br />
1. You must identify coding sequences in your genome<br />
(e.g., using <a href="https://github.com/hyattpd/Prodigal">prodigal</a>)<br />
2. You must annotate a set of highly-expressed genes, typically ribosomal proteins</p>
<p>It is helpful if (in frame) coding sequences are stored in a fasta file for easy loading. Additionally, you will need to have a logical vector describing which genes are part of your highly expressed set. This should become a bit clearer with the example below.</p>
</div>
<div id="a-minimal-example" class="section level2">
<h2>A Minimal Example</h2>
<p>We have included an example genome assembly with this package, downloaded from NCBI’s RefSeq database (<em>Streptococcus pyogenes</em> M1, GCF_000349925.2). This is simply a fasta file with the predicted coding sequences and annotations for this genome (provided by NCBI).</p>
<p>Let’s load this file using the <code>readDNAStringSet()</code> function from the <code>Biostrings</code> package (required for <code>gRodon</code> to work).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(gRodon)
<span class="kw">library</span>(Biostrings)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">path_to_genome &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">'extdata'</span>,
  <span class="st">'GCF_000349925.2_ASM34992v2_cds_from_genomic.fna.gz'</span>,
  <span class="dt">package =</span> <span class="st">'gRodon'</span>)
genes &lt;-<span class="st"> </span><span class="kw">readDNAStringSet</span>(path_to_genome)</code></pre></div>
<p>We also need a set of highly expressed genes. In general, the ribosomal proteins are a good set of genes to use for this purpose. Since these proteins should already be annotated in our example file (try running <code>names(genes)</code> to see the annotations from NCBI) we can use <code>grep()</code> to search for them (specifically, <code>grepl()</code> to return a logical vector).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">highly_expressed &lt;-<span class="st"> </span><span class="kw">grepl</span>(<span class="st">&quot;ribosomal protein&quot;</span>,<span class="kw">names</span>(genes),<span class="dt">ignore.case =</span> T)</code></pre></div>
<p>And now we are ready to predict the growth rate of <em>S. pyogenes</em> M1.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">predictGrowth</span>(genes, highly_expressed)
<span class="co">#&gt; Warning in filterSeq(genes = genes, highly_expressed = highly_expressed, :</span>
<span class="co">#&gt; There were 146 genes either with lengths not multiples of 3 or not above length</span>
<span class="co">#&gt; threshold (default 240bp), these genes have been ignored</span>
<span class="co">#&gt; $CUBHE</span>
<span class="co">#&gt; [1] 0.9148924</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $ConsistencyHE</span>
<span class="co">#&gt; [1] 0.5062119</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $CPB</span>
<span class="co">#&gt; [1] -0.3886505</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $FilteredSequences</span>
<span class="co">#&gt; [1] 146</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $d</span>
<span class="co">#&gt; [1] 0.2987448</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $LowerCI</span>
<span class="co">#&gt; [1] 0.2349466</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $UpperCI</span>
<span class="co">#&gt; [1] 0.3830901</span></code></pre></div>
<p>The warning at the top is letting you know that there were a number of genes in your fasta that either weren’t a length multiple of three, or were under the length threshold of 100bp where estimates of bias will be unreliable. Don’t worry about this unless the number of filtered genes is very high compared to the total number of genes in your file. We expect some genes to typically get filtered.</p>
<p>The output contains several (hopefully) useful quantities:</p>
<ul>
<li><code>$CUBHE</code> is the codon usage bias of the highly expressed genes relative to the genomic background</li>
<li><code>$ConsistencyHE</code> is the codon usage bias of the highly expressed genes relative to the all highly expressed genes</li>
<li><code>$CPB</code> is the genome-wide codon pair bias</li>
<li><code>$FilteredSequences</code> is the number of sequences filtered due to length</li>
<li><code>$d</code> is the estimated doubling time from <code>gRodon</code>. This is, presumably, what you came here for</li>
<li><code>$LowerCI</code> and <code>$UpperCI</code> are the 95% confidence intervals for <code>d</code></li>
</ul>
<p>For the most part, that’s all you need to know. There are a few specific use cases we discuss below that require a slightly more thoughful application of <code>gRodon</code>. These are:</p>
<ul>
<li>The use of partial genomic data (e.g., incomplete SAGs or MAGs)<br />
</li>
<li>The use of metagenomic data<br />
</li>
<li>Predicting growth rates when growth is very slow<br />
</li>
<li>Predicting the growth rate of a psychrophile or thermophile</li>
</ul>
<div id="partial-mode" class="section level3">
<h3>Partial Mode</h3>
<p>What if you don’t have a nice, complete genome … can you still use <code>gRodon</code>? Yes, you can, though you may want to use either “partial” or “metagenome” mode.</p>
<p>When your genome is incomplete, you may have insufficient data to accurately estimate the codon pair bias (since there are many possible codon pairs). In this case you can set <code>gRodon</code> to “partial” mode, which excludes pair-bias from the prediction. This is probably a good choice when working with incomplete SAGs and MAGs. The expected decrease in accuracy is quite small (see original <code>gRodon</code> paper).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">predictGrowth</span>(genes, highly_expressed, <span class="dt">mode=</span><span class="st">&quot;partial&quot;</span>)
<span class="co">#&gt; Warning in filterSeq(genes = genes, highly_expressed = highly_expressed, :</span>
<span class="co">#&gt; There were 146 genes either with lengths not multiples of 3 or not above length</span>
<span class="co">#&gt; threshold (default 240bp), these genes have been ignored</span>
<span class="co">#&gt; $CUBHE</span>
<span class="co">#&gt; [1] 0.9148924</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $ConsistencyHE</span>
<span class="co">#&gt; [1] 0.5062119</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $CPB</span>
<span class="co">#&gt; [1] -0.3886505</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $FilteredSequences</span>
<span class="co">#&gt; [1] 146</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $d</span>
<span class="co">#&gt; [1] 0.3631482</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $LowerCI</span>
<span class="co">#&gt; [1] 0.2840725</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $UpperCI</span>
<span class="co">#&gt; [1] 0.4683562</span></code></pre></div>
</div>
<div id="metagenome-mode" class="section level3">
<h3>Metagenome Mode</h3>
<p>For bulk estimation from metagenomes, our consistency statistic is not appropriate, since different organisms may prefer different codons (even when they have similar bias values). It also doesn’t make much sense to calculate the pair-bias in this scenario. Thus the model for predicting the mean growth rate of a metagenomic sample excludes consistency and pair-bias. Importantly, metagenome mode is expected to be less accurate than the default mode, so only use this mode if you must (i.e., you have a metagenomic sample).</p>
<p>Let’s try this on one of the samples from <a href="https://doi.org/10.7554/eLife.49816">Okie et al.</a>, as discussed in the <code>gRodon</code> manuscript:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">path_to_metagenome &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">'extdata'</span>,
  <span class="st">'ERR2143764_fastp_prokka_scaffolds.ffn.gz'</span>,
  <span class="dt">package =</span> <span class="st">'gRodon'</span>)
genes &lt;-<span class="st"> </span><span class="kw">readDNAStringSet</span>(path_to_metagenome)
highly_expressed &lt;-<span class="st"> </span><span class="kw">grepl</span>(<span class="st">&quot;ribosomal protein&quot;</span>,<span class="kw">names</span>(genes),<span class="dt">ignore.case =</span> T)
<span class="kw">predictGrowth</span>(genes, highly_expressed, <span class="dt">mode =</span> <span class="st">&quot;metagenome&quot;</span>)
<span class="co">#&gt; Warning in predictGrowth(genes, highly_expressed, mode = &quot;metagenome&quot;): Provide</span>
<span class="co">#&gt; depth_of_coverage for your ORFs for a more realistic average community growth</span>
<span class="co">#&gt; rate</span>
<span class="co">#&gt; Warning in filterSeq(genes = genes, highly_expressed = highly_expressed, :</span>
<span class="co">#&gt; There were 8290 genes either with lengths not multiples of 3 or not above length</span>
<span class="co">#&gt; threshold (default 240bp), these genes have been ignored</span>
<span class="co">#&gt; $CUBHE</span>
<span class="co">#&gt; [1] 0.7435671</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $ConsistencyHE</span>
<span class="co">#&gt; [1] 0.6749404</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $CPB</span>
<span class="co">#&gt; [1] -0.2902042</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $FilteredSequences</span>
<span class="co">#&gt; [1] 8290</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $d</span>
<span class="co">#&gt; [1] 1.568009</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $LowerCI</span>
<span class="co">#&gt; [1] 1.321773</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $UpperCI</span>
<span class="co">#&gt; [1] 1.868001</span></code></pre></div>
<p>You may also want to incorporate a correction for the relative abundance of different organisms in this calculation (e.g., weighted metagenome mode) by specifying the <code>depth_of_coverage</code> option. This should lead to more realistic estimates of the community-wide growth rate. To use this correction provide <code>depth_of_coverage</code> as a vector of mean depths of coverage for each gene. It does not matter if these coverages are normalized since we only care about relative coverage.</p>
<p>First let’s take a look at our read depths:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">path_to_coverage &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">'extdata'</span>,
  <span class="st">'ERR2143764_fastp_map2ffn_counts.tsv'</span>,
  <span class="dt">package =</span> <span class="st">'gRodon'</span>)
read_depths &lt;-<span class="st"> </span><span class="kw">read.delim</span>(path_to_coverage,
                            <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)
depths &lt;-<span class="st"> </span>read_depths<span class="op">$</span>meandepth
<span class="co">#Make sure in the correct order</span>
<span class="kw">names</span>(depths) &lt;-<span class="st"> </span>read_depths<span class="op">$</span>X.rname
depth_of_coverage &lt;-<span class="st"> </span>depths[<span class="kw">gsub</span>(<span class="st">&quot; .*&quot;</span>, <span class="st">&quot;&quot;</span>, <span class="kw">names</span>(genes))]
<span class="kw">head</span>(depth_of_coverage)
<span class="co">#&gt; PNFGHFNJ_00001 PNFGHFNJ_00002 PNFGHFNJ_00003 PNFGHFNJ_00004 PNFGHFNJ_00005 </span>
<span class="co">#&gt;        19.3492        35.1459        32.5432        28.4502        27.3811 </span>
<span class="co">#&gt; PNFGHFNJ_00006 </span>
<span class="co">#&gt;        29.4147</span></code></pre></div>
<p>And now let’s run weighted metagenome mode:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">predictGrowth</span>(genes, 
              highly_expressed, 
              <span class="dt">mode =</span> <span class="st">&quot;metagenome&quot;</span>, 
              <span class="dt">depth_of_coverage =</span> depth_of_coverage)
<span class="co">#&gt; Warning in filterSeq(genes = genes, highly_expressed = highly_expressed, :</span>
<span class="co">#&gt; There were 8290 genes either with lengths not multiples of 3 or not above length</span>
<span class="co">#&gt; threshold (default 240bp), these genes have been ignored</span>
<span class="co">#&gt; $CUBHE</span>
<span class="co">#&gt; [1] 0.6999671</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $ConsistencyHE</span>
<span class="co">#&gt; [1] NA</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $CPB</span>
<span class="co">#&gt; [1] NA</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $FilteredSequences</span>
<span class="co">#&gt; [1] 8290</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $d</span>
<span class="co">#&gt; [1] 2.16715</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $LowerCI</span>
<span class="co">#&gt; [1] 1.847755</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $UpperCI</span>
<span class="co">#&gt; [1] 2.55112</span></code></pre></div>
</div>
<div id="slow-growers" class="section level3">
<h3>Slow Growers</h3>
<p>All codon-usage based predictors of growth will underestimate doubling times when doubling times are very long. In other words, these predictors will often indicate that slow-growing microbes grow more quickly than they actually do. See the original paper for a more in-depth discusison, but this is because codon-usage bias tends to plateau at a lower-bound in slow-growers.</p>
<p>What does this mean for you? Well, <code>gRodon</code> can reliably tell you if your microbe grows slowly, but it can’t tell you how slowly. In practice, doubling time predictions appear to be accurate up to 5 hours. If <code>gRodon</code> predicts a doubling time above 5 hours, you can confidently say your microbe grows slowly, but not quite how slowly (doubling times &gt;5hrs tend to be underestimates). In these cases <code>gRodon</code> will even warn you to be careful:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">path_to_genome &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">'extdata'</span>,
  <span class="st">'GCF_003253775.1_ASM325377v1_cds_from_genomic.fna.gz'</span>,
  <span class="dt">package =</span> <span class="st">'gRodon'</span>)
genes &lt;-<span class="st"> </span><span class="kw">readDNAStringSet</span>(path_to_genome)
highly_expressed &lt;-<span class="st"> </span><span class="kw">grepl</span>(<span class="st">&quot;ribosomal protein&quot;</span>,<span class="kw">names</span>(genes),<span class="dt">ignore.case =</span> T)
<span class="kw">predictGrowth</span>(genes, highly_expressed)
<span class="co">#&gt; Warning in filterSeq(genes = genes, highly_expressed = highly_expressed, :</span>
<span class="co">#&gt; There were 780 genes either with lengths not multiples of 3 or not above length</span>
<span class="co">#&gt; threshold (default 240bp), these genes have been ignored</span>
<span class="co">#&gt; Warning in predictGrowth(genes, highly_expressed): Estimated doubling time</span>
<span class="co">#&gt; &gt;5 hours. CUB signal saturates at approx. 5 hrs... gRodon may underestimate</span>
<span class="co">#&gt; doubling times above this range. Consider simply reporting as '&gt;5hrs'. (In other</span>
<span class="co">#&gt; words, this microbe definitely grows slowly, but we can't tell you quite how</span>
<span class="co">#&gt; slowly).</span>
<span class="co">#&gt; $CUBHE</span>
<span class="co">#&gt; [1] 0.5980514</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $ConsistencyHE</span>
<span class="co">#&gt; [1] 0.5630458</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $CPB</span>
<span class="co">#&gt; [1] -0.3347875</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $FilteredSequences</span>
<span class="co">#&gt; [1] 780</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $d</span>
<span class="co">#&gt; [1] 19.21814</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $LowerCI</span>
<span class="co">#&gt; [1] 13.00419</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $UpperCI</span>
<span class="co">#&gt; [1] 29.05757</span></code></pre></div>
<p>The example above is for a <em>Mycobacterium leprae</em> genome. This organism has a doubling time around 240 hours. So we are off by an order of magnitude, but still well within the “slow grower” range.</p>
</div>
<div id="psychrophiles-and-thermophiles" class="section level3">
<h3>Psychrophiles and Thermophiles</h3>
<p><a href="https://doi.org/10.1371/journal.pgen.1000808">Viera-Silva et al.</a> showed that codon-usage based growth estimators are sensitive to growth temperature. Thermophiles and psychrophiles showed consistently higher/lower growth rates respectively. We included a temperature correstion factor in <code>gRodon</code> to account for this. We caution that this correction is based on only 31 extremophile species, and that <code>gRodon</code> has primarily been validated on mesophilic organisms.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">predictGrowth</span>(genes, highly_expressed, <span class="dt">temperature =</span> <span class="dv">33</span>)
<span class="co">#&gt; Warning in filterSeq(genes = genes, highly_expressed = highly_expressed, :</span>
<span class="co">#&gt; There were 780 genes either with lengths not multiples of 3 or not above length</span>
<span class="co">#&gt; threshold (default 240bp), these genes have been ignored</span>
<span class="co">#&gt; Warning in predictGrowth(genes, highly_expressed, temperature = 33): Estimated</span>
<span class="co">#&gt; doubling time &gt;5 hours. CUB signal saturates at approx. 5 hrs... gRodon may</span>
<span class="co">#&gt; underestimate doubling times above this range. Consider simply reporting as</span>
<span class="co">#&gt; '&gt;5hrs'. (In other words, this microbe definitely grows slowly, but we can't</span>
<span class="co">#&gt; tell you quite how slowly).</span>
<span class="co">#&gt; $CUBHE</span>
<span class="co">#&gt; [1] 0.5980514</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $ConsistencyHE</span>
<span class="co">#&gt; [1] 0.5630458</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $CPB</span>
<span class="co">#&gt; [1] -0.3347875</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $FilteredSequences</span>
<span class="co">#&gt; [1] 780</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $OGT</span>
<span class="co">#&gt; [1] 33</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $d</span>
<span class="co">#&gt; [1] 18.26009</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $LowerCI</span>
<span class="co">#&gt; [1] 12.66798</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $UpperCI</span>
<span class="co">#&gt; [1] 26.85091</span></code></pre></div>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
