---
title: "Predicting maximal growth rates with `gRodon`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{gRodon-vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

The `gRodon` package allows you to predict the maximal growth rate of
a prokaryotic organism from genomic data. For details of how this
is done please see the original paper. Briefly, gRodon exploits codon usage
statistics to detect optimization of highly-expressed genes, an indicator of
rapid growth. We will provide some guidance about when these predictions are 
appropriate below. 

The `gRodon` package is quite simple, with only a single function currently
available to users: `predictGrowth()`. We provide a simple, easy to
use interface that allows any user with a genome in-hand to predict maximal
growth rate. Note that we will use "maximal growth rate", "minimal doubling 
time", and "growth rate" interchangeably below when talking about the output
from `gRodon`. 

The `gRodon` approach is largely based on an earlier program, `growthpred`,
developed by [Viera-Silva et al.](https://doi.org/10.1371/journal.pgen.1000808). 
`gRodon` is intended to be:  
1. More user-friendly (especially for R users)  
2. More accurate (due to the incorporation of additional measures of codon usage)  
3. More accurate on mixed-organism samples (i.e., metagenomes; due to a correction factor for the relative abundance of different organisms)

## Before You Begin

There are two things that `gRodon` does *not* do that you *must* do ahead of 
time yourself in order to predict growth rate:  
1. You must identify coding sequences in your genome  
(e.g., using [prodigal](https://github.com/hyattpd/Prodigal))  
2. You must annotate a set of highly-expressed genes, typically ribosomal 
proteins  

It is helpful if (in frame) coding sequences are stored in a fasta file for 
easy loading. Additionally, you will need to have a logical vector describing
which genes are part of your highly expressed set. This should become a bit
clearer with the example below.

## A Minimal Example

We have included an example genome assembly with this package, downloaded 
from NCBI's RefSeq database (*Streptococcus pyogenes* M1, GCF_000349925.2). This
is simply a fasta file with the predicted coding sequences and 
annotations for this genome (provided by NCBI).

Let's load this file using the `readDNAStringSet()` function from the
`Biostrings` package (required for `gRodon` to work).

```{r setup, message=FALSE}
library(gRodon)
library(Biostrings)
```

```{r load}
path_to_genome <- system.file('extdata',
  'GCF_000349925.2_ASM34992v2_cds_from_genomic.fna.gz',
  package = 'gRodon')
genes <- readDNAStringSet(path_to_genome)
```


We also need a set of highly expressed genes. In general, the ribosomal
proteins are a good set of genes to use for this purpose. Since these proteins
should already be annotated in our example file (try running `names(genes)` to
see the annotations from NCBI) we  can use `grep()` to search for them 
(specifically, `grepl()` to return a logical vector).

```{r HE}
highly_expressed <- grepl("ribosomal protein",names(genes),ignore.case = T)
```
And now we are ready to predict the growth rate of *S. pyogenes* M1.

```{r Pred}
predictGrowth(genes, highly_expressed)
```

The warning at the top is letting you know that there were a number of genes
in your fasta that either weren't a length multiple of three, or were under the 
length threshold of 100bp where estimates of bias will be unreliable. Don't
worry about this unless the number of filtered genes is very high compared
to the total number of genes in your file. We expect some genes to typically get
filtered.

The output contains several (hopefully) useful quantities:

*  `$CUBHE` is the codon usage bias of the highly expressed genes relative to
the genomic background
*  `$ConsistencyHE` is the codon usage bias of the highly expressed genes relative to
the all highly expressed genes
*  `$CPB` is the genome-wide codon pair bias
*  `$FilteredSequences` is the number of sequences filtered due to length
*  `$d` is the estimated doubling time from `gRodon`. This is, presumably, what
you came here for
*  `$LowerCI` and `$UpperCI` are the 95% confidence intervals for `d`

For the most part, that's all you need to know. There are a few specific use
cases we discuss below that require a slightly more thoughful application of
`gRodon`. These are:

*  The use of partial genomic data (e.g., incomplete SAGs or MAGs)  
*  The use of metagenomic data  
*  Predicting growth rates when growth is very slow  
*  Predicting the growth rate of a psychrophile or thermophile  

### Partial Mode

What if you don't have a nice, complete genome ... can you still use `gRodon`? 
Yes, you can, though you may want to use either "partial" or "metagenome" mode.

When your genome is incomplete, you may have insufficient data to accurately
estimate the codon pair bias (since there are many possible codon pairs). In 
this case you can set `gRodon` to "partial" mode, which excludes pair-bias 
from the prediction. This is probably a good choice when working with incomplete
SAGs and MAGs. The expected decrease in accuracy is quite small (see original 
`gRodon` paper).

```{r PredP}
predictGrowth(genes, highly_expressed, mode="partial")
```

### Metagenome Mode

For bulk estimation from metagenomes, 
our consistency statistic is not appropriate, since different
organisms may prefer different codons (even when they have similar bias values).
It also doesn't make much sense to calculate the pair-bias in this scenario.
Thus the model for predicting the mean growth rate of a metagenomic sample
excludes consistency and pair-bias. Importantly, metagenome mode is expected to
be less accurate than the default mode, so only use this mode if you must 
(i.e., you have a metagenomic sample).

Let's try this on one of the samples from [Okie et al.](https://doi.org/10.7554/eLife.49816), as discussed in the
`gRodon` manuscript:

```{r PredM}
path_to_metagenome <- system.file('extdata',
  'ERR2143764_fastp_prokka_scaffolds.ffn.gz',
  package = 'gRodon')
genes <- readDNAStringSet(path_to_metagenome)
highly_expressed <- grepl("ribosomal protein",names(genes),ignore.case = T)
predictGrowth(genes, highly_expressed, mode = "metagenome")
```

You may also want to incorporate a correction for the relative abundance of different
organisms in this calculation (e.g., weighted metagenome mode) by specifying the `depth_of_coverage`
option. This should lead to more realistic estimates of the community-wide growth rate.
To use this correction provide `depth_of_coverage` as a vector of mean depths of coverage for each gene.
It does not matter if these coverages are normalized since we only care about relative coverage.

First let's take a look at our read depths:

```{r readdepths}
path_to_coverage <- system.file('extdata',
  'ERR2143764_fastp_map2ffn_counts.tsv',
  package = 'gRodon')
read_depths <- read.delim(path_to_coverage,
                            stringsAsFactors = FALSE)
depths <- read_depths$meandepth
#Make sure in the correct order
names(depths) <- read_depths$X.rname
depth_of_coverage <- depths[gsub(" .*", "", names(genes))]
head(depth_of_coverage)
```

And now let's run weighted metagenome mode:

```{r weightedM}
predictGrowth(genes, 
              highly_expressed, 
              mode = "metagenome", 
              depth_of_coverage = depth_of_coverage)
```

### Slow Growers

All codon-usage based predictors of growth will underestimate doubling times
when doubling times are very long. In other words, these predictors will often 
indicate that slow-growing microbes grow more quickly than they actually do. 
See the original paper for a more in-depth discusison, but this is because 
codon-usage bias tends to plateau at a lower-bound in slow-growers. 

What does this mean for you? Well, `gRodon` can reliably tell you if your
microbe grows slowly, but it can't tell you how slowly. In practice, doubling 
time predictions appear to be accurate up to 5 hours. If `gRodon` predicts a
doubling time above 5 hours, you can confidently say your microbe grows slowly,
but not quite how slowly (doubling times >5hrs tend to be underestimates). 
In these cases `gRodon` will even warn you to be  careful:

```{r slow}
path_to_genome <- system.file('extdata',
  'GCF_003253775.1_ASM325377v1_cds_from_genomic.fna.gz',
  package = 'gRodon')
genes <- readDNAStringSet(path_to_genome)
highly_expressed <- grepl("ribosomal protein",names(genes),ignore.case = T)
predictGrowth(genes, highly_expressed)
```

The example above is for a *Mycobacterium leprae* genome. This organism has a 
doubling time around 240 hours. So we are off by an order of magnitude, but 
still well within the "slow grower" range.

### Psychrophiles and Thermophiles

[Viera-Silva et al.](https://doi.org/10.1371/journal.pgen.1000808) showed that
codon-usage based growth estimators are sensitive to growth temperature. 
Thermophiles and psychrophiles showed consistently higher/lower growth rates 
respectively. We included a temperature correstion factor in `gRodon` to account
for this. We caution that this correction is based on only 31 extremophile
species, and that `gRodon` has primarily been validated on 
mesophilic organisms.

```{r temp}
predictGrowth(genes, highly_expressed, temperature = 33)
```
